File: builder/MrrBuilder.java
/*
 * ╔══════════════════════════════════════════════════════════════════╗
 * ║  MRR Builder — Java Build Tool v1.0.1                          ║
 * ║                                                                  ║
 * ║  MRR projelerini .mbuild formatında paketler.                  ║
 * ║  .mbuild = ZIP tabanlı self-contained çalıştırılabilir paket.  ║
 * ║                                                                  ║
 * ║  Kullanım:                                                      ║
 * ║    javac MrrBuilder.java && java MrrBuilder <dosya.mrr>        ║
 * ║                                                                  ║
 * ║  .mbuild içeriği:                                               ║
 * ║    manifest.json  — Paket metadata                              ║
 * ║    source/        — MRR kaynak dosyaları                        ║
 * ║    runtime/       — MRR interpreter                             ║
 * ║    entry.py       — Başlatıcı script                            ║
 * ╚══════════════════════════════════════════════════════════════════╝
 */

import java.io.*;
import java.nio.file.*;
import java.nio.file.attribute.BasicFileAttributes;
import java.util.*;
import java.util.zip.*;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;

public class MrrBuilder {

    private static final String VERSION = "1.0.1";
    private static final String CODENAME = "Resilient";

    // ─────────────────────────────────────────────────────
    // Ana Giriş Noktası
    // ─────────────────────────────────────────────────────

    public static void main(String[] args) {
        if (args.length < 1) {
            printUsage();
            System.exit(1);
        }

        String sourceFile = args[0];
        String format = "mb"; // Varsayılan: .mbuild
        String outputFile = null;

        // Argümanları ayrıştır
        for (int i = 1; i < args.length; i++) {
            switch (args[i]) {
                case "--format":
                case "-f":
                    if (i + 1 < args.length) {
                        format = args[++i];
                    }
                    break;
                case "--output":
                case "-o":
                    if (i + 1 < args.length) {
                        outputFile = args[++i];
                    }
                    break;
                case "--help":
                case "-h":
                    printUsage();
                    System.exit(0);
                    break;
            }
        }

        MrrBuilder builder = new MrrBuilder();
        
        try {
            if ("mb".equals(format)) {
                builder.buildMbuild(sourceFile, outputFile);
            } else if ("exe".equals(format)) {
                System.out.println("[BİLGİ] EXE build için C# MrrBuilder kullanın:");
                System.out.println("  dotnet run MrrBuilder.cs " + sourceFile + " --format exe");
                System.out.println("\n  Veya Python builder: mrr build " + sourceFile + " --format exe");
            } else {
                System.err.println("[HATA] Bilinmeyen format: " + format);
                System.exit(1);
            }
        } catch (Exception e) {
            System.err.println("[HATA] Build başarısız: " + e.getMessage());
            System.exit(1);
        }
    }

    // ─────────────────────────────────────────────────────
    // .mbuild Build
    // ─────────────────────────────────────────────────────

    public void buildMbuild(String sourceFile, String outputFile) throws IOException {
        Path sourcePath = Paths.get(sourceFile).toAbsolutePath();
        
        if (!Files.exists(sourcePath)) {
            throw new FileNotFoundException("Dosya bulunamadı: " + sourceFile);
        }

        String projectName = sourcePath.getFileName().toString().replace(".mrr", "");
        if (outputFile == null) {
            outputFile = projectName + ".mbuild";
        }

        printBanner("mb");
        System.out.println("  Kaynak:  " + sourcePath);
        System.out.println("  Çıktı:   " + outputFile);
        System.out.println("  Format:  .mbuild (self-contained)");
        System.out.println();

        Path outputPath = sourcePath.getParent().resolve(outputFile);
        Path tempDir = Files.createTempDirectory("mrr_build_");

        try {
            // 1. Manifest oluştur
            System.out.println("  [1/4] Manifest oluşturuluyor...");
            createManifest(tempDir, projectName, sourcePath);

            // 2. Kaynak dosyaları kopyala
            System.out.println("  [2/4] Kaynak dosyalar kopyalanıyor...");
            Path sourceDir = tempDir.resolve("source");
            Files.createDirectories(sourceDir);
            Files.copy(sourcePath, sourceDir.resolve(sourcePath.getFileName()));
            
            // Aynı dizindeki diğer .mrr dosyalarını da ekle
            try (DirectoryStream<Path> stream = Files.newDirectoryStream(sourcePath.getParent(), "*.mrr")) {
                for (Path mrrFile : stream) {
                    if (!mrrFile.equals(sourcePath)) {
                        Files.copy(mrrFile, sourceDir.resolve(mrrFile.getFileName()));
                    }
                }
            }

            // 3. Runtime kopyala
            System.out.println("  [3/4] MRR Runtime paketleniyor...");
            Path runtimeDir = tempDir.resolve("runtime");
            Files.createDirectories(runtimeDir);
            
            // MRR root dizinini bul (builder'ın bulunduğu dizinin üst dizini)
            Path mrrRoot = findMrrRoot(sourcePath);
            if (mrrRoot != null) {
                Path interpDir = mrrRoot.resolve("interpreter");
                if (Files.exists(interpDir)) {
                    copyDirectory(interpDir, runtimeDir.resolve("interpreter"));
                }
                Path stdlibDir = mrrRoot.resolve("stdlib");
                if (Files.exists(stdlibDir)) {
                    copyDirectory(stdlibDir, runtimeDir.resolve("stdlib"));
                }
            }

            // Entry point oluştur
            createEntryPoint(tempDir, sourcePath.getFileName().toString());

            // 4. ZIP olarak paketle
            System.out.println("  [4/4] .mbuild arşivi oluşturuluyor...");
            createZipArchive(tempDir, outputPath);

            long sizeKb = Files.size(outputPath) / 1024;
            System.out.println("\n  \033[32m✓ Build başarılı!\033[0m");
            System.out.println("    Dosya: " + outputPath);
            System.out.println("    Boyut: " + sizeKb + " KB");

        } finally {
            // Geçici dizini temizle
            deleteDirectory(tempDir);
        }
    }

    // ─────────────────────────────────────────────────────
    // Yardımcı Metodlar
    // ─────────────────────────────────────────────────────

    private void createManifest(Path dir, String projectName, Path sourcePath) throws IOException {
        String manifest = String.format(
            "{\n" +
            "  \"name\": \"%s\",\n" +
            "  \"version\": \"%s\",\n" +
            "  \"entry\": \"source/%s\",\n" +
            "  \"runtime\": \"mrr\",\n" +
            "  \"runtime_version\": \"%s\",\n" +
            "  \"build_date\": \"%s\",\n" +
            "  \"builder\": \"MrrBuilder.java\",\n" +
            "  \"platform\": \"%s\",\n" +
            "  \"format\": \"mbuild\",\n" +
            "  \"format_version\": \"1.0\"\n" +
            "}",
            projectName,
            VERSION,
            sourcePath.getFileName(),
            VERSION,
            LocalDateTime.now().format(DateTimeFormatter.ISO_LOCAL_DATE_TIME),
            System.getProperty("os.name")
        );
        Files.writeString(dir.resolve("manifest.json"), manifest);
    }

    private void createEntryPoint(Path dir, String sourceFileName) throws IOException {
        String entry = String.format(
            "#!/usr/bin/env python3\n" +
            "\"\"\"MRR .mbuild Entry Point — Auto-generated by MrrBuilder.java v%s\"\"\"\n" +
            "import sys, os\n" +
            "runtime_dir = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'runtime')\n" +
            "sys.path.insert(0, runtime_dir)\n" +
            "source_file = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'source', '%s')\n" +
            "from interpreter.mrr_lexer import Lexer\n" +
            "from interpreter.mrr_parser import Parser\n" +
            "from interpreter.mrr_evaluator import Evaluator\n" +
            "from interpreter.mrr_ffi import FFIBridge, SandboxPolicy\n" +
            "with open(source_file, 'r', encoding='utf-8') as f:\n" +
            "    code = f.read()\n" +
            "lexer = Lexer(code, source_file)\n" +
            "tokens = lexer.tokenize()\n" +
            "parser = Parser(tokens)\n" +
            "ast = parser.parse()\n" +
            "ffi = FFIBridge(policy=SandboxPolicy())\n" +
            "evaluator = Evaluator()\n" +
            "evaluator.ffi_bridge = ffi\n" +
            "evaluator.execute(ast)\n",
            VERSION, sourceFileName
        );
        Files.writeString(dir.resolve("entry.py"), entry);
    }

    private Path findMrrRoot(Path fromPath) {
        Path current = fromPath.getParent();
        for (int i = 0; i < 5; i++) {
            if (current == null) break;
            if (Files.exists(current.resolve("interpreter")) &&
                Files.exists(current.resolve("stdlib"))) {
                return current;
            }
            current = current.getParent();
        }
        return null;
    }

    private void copyDirectory(Path source, Path target) throws IOException {
        Files.walkFileTree(source, new SimpleFileVisitor<Path>() {
            @Override
            public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) throws IOException {
                String dirName = dir.getFileName().toString();
                if (dirName.equals("__pycache__") || dirName.equals(".git")) {
                    return FileVisitResult.SKIP_SUBTREE;
                }
                Files.createDirectories(target.resolve(source.relativize(dir)));
                return FileVisitResult.CONTINUE;
            }

            @Override
            public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
                String fileName = file.getFileName().toString();
                if (fileName.endsWith(".pyc") || fileName.endsWith(".pyo")) {
                    return FileVisitResult.CONTINUE;
                }
                Files.copy(file, target.resolve(source.relativize(file)),
                           StandardCopyOption.REPLACE_EXISTING);
                return FileVisitResult.CONTINUE;
            }
        });
    }

    private void createZipArchive(Path sourceDir, Path outputFile) throws IOException {
        try (ZipOutputStream zos = new ZipOutputStream(
                new BufferedOutputStream(new FileOutputStream(outputFile.toFile())))) {
            zos.setLevel(Deflater.BEST_COMPRESSION);
            
            Files.walkFileTree(sourceDir, new SimpleFileVisitor<Path>() {
                @Override
                public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
                    String entryName = sourceDir.relativize(file).toString().replace("\\", "/");
                    zos.putNextEntry(new ZipEntry(entryName));
                    Files.copy(file, zos);
                    zos.closeEntry();
                    return FileVisitResult.CONTINUE;
                }
            });
        }
    }

    private void deleteDirectory(Path dir) {
        try {
            Files.walkFileTree(dir, new SimpleFileVisitor<Path>() {
                @Override
                public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
                    Files.delete(file);
                    return FileVisitResult.CONTINUE;
                }

                @Override
                public FileVisitResult postVisitDirectory(Path d, IOException exc) throws IOException {
                    Files.delete(d);
                    return FileVisitResult.CONTINUE;
                }
            });
        } catch (IOException ignored) {}
    }

    // ─────────────────────────────────────────────────────
    // UI
    // ─────────────────────────────────────────────────────

    private void printBanner(String mode) {
        String modeStr = "mb".equals(mode) ? ".mbuild Mode" : "EXE Mode    ";
        System.out.println();
        System.out.println("╔══════════════════════════════════════════╗");
        System.out.println("║  MRR Builder v" + VERSION + " — " + modeStr + "   ║");
        System.out.println("╚══════════════════════════════════════════╝");
        System.out.println();
    }

    private static void printUsage() {
        System.out.println();
        System.out.println("╔══════════════════════════════════════════════╗");
        System.out.println("║  MRR Builder v" + VERSION + " (Java)                 ║");
        System.out.println("║  .mbuild Paket Oluşturma Aracı              ║");
        System.out.println("╚══════════════════════════════════════════════╝");
        System.out.println();
        System.out.println("  Kullanım:");
        System.out.println("    java MrrBuilder <dosya.mrr> [seçenekler]");
        System.out.println();
        System.out.println("  Seçenekler:");
        System.out.println("    --format, -f <mb|exe>   Build formatı (varsayılan: mb)");
        System.out.println("    --output, -o <dosya>    Çıktı dosya adı");
        System.out.println("    --help, -h              Bu yardım mesajı");
        System.out.println();
        System.out.println("  Örnekler:");
        System.out.println("    java MrrBuilder proje.mrr");
        System.out.println("    java MrrBuilder proje.mrr --format mb -o output.mbuild");
        System.out.println();
    }
}

File: builder/MrrBuilder.cs
/*
 * ╔══════════════════════════════════════════════════════════════════╗
 * ║  MRR Builder — C# EXE Build Tool v1.0.1                       ║
 * ║                                                                  ║
 * ║  MRR projelerini native Windows .exe dosyasına dönüştürür.     ║
 * ║  .mbuild paketini self-extracting executable olarak paketler.  ║
 * ║                                                                  ║
 * ║  Kullanım:                                                      ║
 * ║    dotnet run MrrBuilder.cs <dosya.mrr> --format exe            ║
 * ║    csc MrrBuilder.cs && MrrBuilder.exe <dosya.mrr>              ║
 * ╚══════════════════════════════════════════════════════════════════╝
 */

using System;
using System.IO;
using System.IO.Compression;
using System.Text;
using System.Diagnostics;
using System.Runtime.InteropServices;

namespace MRR.Builder
{
    /// <summary>
    /// MRR C# Builder — Native .exe oluşturma aracı.
    /// 
    /// İki mod destekler:
    ///   1. Self-Extracting EXE: .mbuild paketini EXE'ye gömülü olarak paketler
    ///   2. PyInstaller Wrapper: Python PyInstaller üzerinden EXE oluşturur
    /// </summary>
    class MrrBuilder
    {
        private const string VERSION = "1.0.1";
        private const string CODENAME = "Resilient";

        static void Main(string[] args)
        {
            if (args.Length < 1)
            {
                PrintUsage();
                return;
            }

            string sourceFile = args[0];
            string outputFile = null;
            string format = "exe";

            for (int i = 1; i < args.Length; i++)
            {
                switch (args[i])
                {
                    case "--output":
                    case "-o":
                        if (i + 1 < args.Length) outputFile = args[++i];
                        break;
                    case "--format":
                    case "-f":
                        if (i + 1 < args.Length) format = args[++i];
                        break;
                    case "--help":
                    case "-h":
                        PrintUsage();
                        return;
                }
            }

            var builder = new MrrBuilder();

            try
            {
                if (format == "exe")
                    builder.BuildExe(sourceFile, outputFile);
                else if (format == "mb")
                {
                    Console.WriteLine("[BİLGİ] .mbuild build için Java MrrBuilder kullanın:");
                    Console.WriteLine($"  java MrrBuilder {sourceFile} --format mb");
                }
            }
            catch (Exception ex)
            {
                Console.ForegroundColor = ConsoleColor.Red;
                Console.WriteLine($"[HATA] Build başarısız: {ex.Message}");
                Console.ResetColor();
                Environment.Exit(1);
            }
        }

        // ─────────────────────────────────────────────────────
        // EXE Build
        // ─────────────────────────────────────────────────────

        /// <summary>
        /// MRR dosyasını native .exe'ye dönüştürür.
        /// Strateji: 
        ///   1. MRR kaynağını ve runtime'ı bir .mbuild'e paketle
        ///   2. Self-extracting launcher oluştur
        ///   3. .mbuild'i EXE'ye gömülü kaynak olarak ekle
        /// </summary>
        void BuildExe(string sourceFile, string outputFile)
        {
            string fullPath = Path.GetFullPath(sourceFile);

            if (!File.Exists(fullPath))
                throw new FileNotFoundException($"Dosya bulunamadı: {sourceFile}");

            string projectName = Path.GetFileNameWithoutExtension(fullPath);
            outputFile ??= projectName + ".exe";

            PrintBanner();
            Console.WriteLine($"  Kaynak:  {fullPath}");
            Console.WriteLine($"  Çıktı:   {outputFile}");
            Console.WriteLine($"  Format:  .exe (native Windows)");
            Console.WriteLine();

            // 1. Önce .mbuild oluştur
            Console.WriteLine("  [1/3] MRR paketi hazırlanıyor...");
            string mbuildPath = Path.Combine(
                Path.GetDirectoryName(fullPath)!,
                projectName + ".mbuild"
            );

            CreateMbuild(fullPath, mbuildPath, projectName);

            // 2. Self-extracting launcher oluştur
            Console.WriteLine("  [2/3] EXE launcher oluşturuluyor...");
            string launcherSource = GenerateLauncherSource(projectName, fullPath);

            // 3. Compile
            Console.WriteLine("  [3/3] Derleniyor...");
            
            // Python/PyInstaller yaklaşımı kullan
            string wrapperPy = Path.Combine(
                Path.GetDirectoryName(fullPath)!,
                $"_mrr_exe_{projectName}.py"
            );
            File.WriteAllText(wrapperPy, launcherSource);

            bool success = RunPyInstaller(wrapperPy, projectName, Path.GetDirectoryName(fullPath)!);

            // Temizlik
            if (File.Exists(mbuildPath)) File.Delete(mbuildPath);
            if (File.Exists(wrapperPy)) File.Delete(wrapperPy);

            if (success)
            {
                string exePath = Path.Combine(Path.GetDirectoryName(fullPath)!, outputFile);
                if (File.Exists(exePath))
                {
                    var info = new FileInfo(exePath);
                    Console.ForegroundColor = ConsoleColor.Green;
                    Console.WriteLine($"\n  ✓ EXE build başarılı!");
                    Console.ResetColor();
                    Console.WriteLine($"    Dosya: {exePath}");
                    Console.WriteLine($"    Boyut: {info.Length / (1024.0 * 1024.0):F1} MB");
                }
            }
            else
            {
                Console.ForegroundColor = ConsoleColor.Yellow;
                Console.WriteLine("  ⚠ PyInstaller bulunamadı veya hata oluştu.");
                Console.WriteLine("  Yüklemek için: pip install pyinstaller");
                Console.ResetColor();
            }
        }

        void CreateMbuild(string sourceFile, string outputPath, string projectName)
        {
            if (File.Exists(outputPath)) File.Delete(outputPath);

            using var zip = ZipFile.Open(outputPath, ZipArchiveMode.Create);

            // Manifest
            string manifest = $@"{{
  ""name"": ""{projectName}"",
  ""version"": ""{VERSION}"",
  ""entry"": ""source/{Path.GetFileName(sourceFile)}"",
  ""runtime"": ""mrr"",
  ""runtime_version"": ""{VERSION}"",
  ""build_date"": ""{DateTime.Now:O}"",
  ""builder"": ""MrrBuilder.cs"",
  ""platform"": ""{RuntimeInformation.OSDescription}"",
  ""format"": ""mbuild"",
  ""format_version"": ""1.0""
}}";
            var manifestEntry = zip.CreateEntry("manifest.json");
            using (var writer = new StreamWriter(manifestEntry.Open()))
                writer.Write(manifest);

            // Source
            var sourceEntry = zip.CreateEntry($"source/{Path.GetFileName(sourceFile)}");
            using (var writer = new StreamWriter(sourceEntry.Open()))
                writer.Write(File.ReadAllText(sourceFile));
        }

        string GenerateLauncherSource(string projectName, string sourceFile)
        {
            string sourceCode = File.ReadAllText(sourceFile)
                .Replace("\\", "\\\\")
                .Replace("\"\"\"", "\\\"\\\"\\\"");

            return $@"#!/usr/bin/env python3
""""""MRR EXE — Auto-generated by MrrBuilder.cs v{VERSION}""""""
import sys, os

# MRR root'u bul
mrr_root = os.environ.get('MRR_ROOT', '')
if mrr_root:
    sys.path.insert(0, mrr_root)

# Gömülü kaynak
SOURCE = """"""{sourceCode}""""""

try:
    from interpreter.mrr_lexer import Lexer
    from interpreter.mrr_parser import Parser
    from interpreter.mrr_evaluator import Evaluator
    from interpreter.mrr_ffi import FFIBridge, SandboxPolicy

    lexer = Lexer(SOURCE, '{Path.GetFileName(sourceFile)}')
    tokens = lexer.tokenize()
    parser = Parser(tokens)
    ast = parser.parse()
    ffi = FFIBridge(policy=SandboxPolicy())
    evaluator = Evaluator()
    evaluator.ffi_bridge = ffi
    evaluator.execute(ast)
except ImportError:
    print('MRR runtime bulunamadı. MRR_ROOT ortam değişkenini ayarlayın.')
    sys.exit(1)
";
        }

        bool RunPyInstaller(string scriptPath, string projectName, string outputDir)
        {
            try
            {
                string iconPath = FindIcon();
                var psi = new ProcessStartInfo
                {
                    FileName = "python",
                    Arguments = $"-m PyInstaller --onefile --name {projectName} " +
                               $"--distpath \"{outputDir}\" " +
                               $"--specpath \"{outputDir}\" " +
                               $"--workpath \"{Path.Combine(outputDir, "build")}\" " +
                               (iconPath != null ? $"--icon=\"{iconPath}\" " : "") +
                               $"--clean \"{scriptPath}\"",
                    RedirectStandardOutput = true,
                    RedirectStandardError = true,
                    UseShellExecute = false,
                    CreateNoWindow = true
                };

                using var process = Process.Start(psi);
                process?.WaitForExit(120000); // 2 dakika timeout

                // Temizlik
                string buildDir = Path.Combine(outputDir, "build");
                if (Directory.Exists(buildDir)) Directory.Delete(buildDir, true);
                string specFile = Path.Combine(outputDir, $"{projectName}.spec");
                if (File.Exists(specFile)) File.Delete(specFile);

                return process?.ExitCode == 0;
            }
            catch
            {
                return false;
            }
        }

        string? FindIcon()
        {
            // Proje kökünde assets/mrr.ico ara
            string? dir = Path.GetDirectoryName(
                System.Reflection.Assembly.GetExecutingAssembly().Location);
            
            for (int i = 0; i < 4 && dir != null; i++)
            {
                string iconPath = Path.Combine(dir, "assets", "mrr.ico");
                if (File.Exists(iconPath)) return iconPath;
                dir = Path.GetDirectoryName(dir);
            }
            return null;
        }

        // ─────────────────────────────────────────────────────
        // UI
        // ─────────────────────────────────────────────────────

        void PrintBanner()
        {
            Console.WriteLine();
            Console.WriteLine("╔══════════════════════════════════════════╗");
            Console.WriteLine($"║  MRR Builder v{VERSION} — C# EXE Mode     ║");
            Console.WriteLine("╚══════════════════════════════════════════╝");
            Console.WriteLine();
        }

        static void PrintUsage()
        {
            Console.WriteLine();
            Console.WriteLine("╔══════════════════════════════════════════════╗");
            Console.WriteLine($"║  MRR Builder v{VERSION} (C#)                  ║");
            Console.WriteLine("║  Native EXE Oluşturma Aracı                 ║");
            Console.WriteLine("╚══════════════════════════════════════════════╝");
            Console.WriteLine();
            Console.WriteLine("  Kullanım:");
            Console.WriteLine("    MrrBuilder.exe <dosya.mrr> [seçenekler]");
            Console.WriteLine();
            Console.WriteLine("  Seçenekler:");
            Console.WriteLine("    --output, -o <dosya>   Çıktı dosya adı");
            Console.WriteLine("    --help, -h             Bu yardım mesajı");
            Console.WriteLine();
            Console.WriteLine("  Örnekler:");
            Console.WriteLine("    MrrBuilder.exe proje.mrr");
            Console.WriteLine("    MrrBuilder.exe proje.mrr -o output.exe");
            Console.WriteLine();
        }
    }
}
